spec/controllers/api/orders_controller_spec.rb

```diff
@@ -156,6 +158,109 @@ module Api
+          get :show, id: order.number
+          expect_order
+        end
+
+        it "returns unauthorized, as the order product's supplier owner" do
+          allow(controller).to receive(:spree_current_user) { order.line_items.first.variant.product.supplier.owner }
+          get :show, id: order.number
+          assert_unauthorized!
+        end
+
+        it "returns the order, as the Order Cycle coorinator owner" do
+          allow(controller).to receive(:spree_current_user) { order.order_cycle.coordinator.owner }
+          get :show, id: order.number
+          expect_order
+        end
+      end
+
+      context "as distributor owner" do
+        let!(:order) { create(:completed_order_with_fees, order_cycle: order_cycle, distributor: distributor ) }
+
+        before { allow(controller).to receive(:spree_current_user) { order.distributor.owner } }
+
+        it "can view an order not in a standard state" do
+          order.update_attributes(completed_at: nil, state: 'shipped')
+          get :show, id: order.number
+          expect_order
+        end
+
+        it "returns an order with all required fields" do
+          get :show, id: order.number
+          expect_order
+          expect_detailed_attributes_to_be_present(json_response)
+
+          expect(json_response[:bill_address][:address1]).to eq(order.bill_address.address1)
+          expect(json_response[:bill_address][:lastname]).to eq(order.bill_address.lastname)
+          expect(json_response[:ship_address][:address1]).to eq(order.ship_address.address1)
+          expect(json_response[:ship_address][:lastname]).to eq(order.ship_address.lastname)
+          expect(json_response[:shipping_method][:name]).to eq(order.shipping_method.name)
+          json_response[:adjustments].each do |adjustment|
+            if adjustment[:label] == "Transaction fee"
+              expect(adjustment[:amount]).to eq(order.adjustments.payment_fee.first.amount.to_s)
+            elsif json_response[:adjustments].first[:label] == "Shipping"
+              expect(adjustment[:amount]).to eq(order.adjustments.shipping.first.amount.to_s)
+            end
+          end
+          expect(json_response[:payments].first[:amount]).to eq(order.payments.first.amount.to_s)
+          expect(json_response[:line_items].size).to eq(order.line_items.size)
+          expect(json_response[:line_items].first[:variant][:product_name]). to eq(order.line_items.first.variant.product.name)
+        end
+      end
+
+      def expect_order
+        expect(response.status).to eq 200
+        expect_correct_order(json_response, order)
+      end
+
+      def expect_correct_order(json_response, order)
+        expect(json_response[:number]).to eq(order.number)
+        expect(json_response[:email]).to eq(order.email)
+      end
+
+      def expect_detailed_attributes_to_be_present(json_response)
+        expect(order_detailed_attributes.all?{ |attr| json_response.key? attr.to_s }).to eq(true)
```

Comment:
I think either [include](https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/include-matcher) or [have_attributes](https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/have-attributes-matcher) RSpec matchers already implement this.

I usually check this with `expect(order.attributes).to include(expected_attributes_hash)` so less code of our own to maintain.

Reply:
yeah, I tried these in previous efforts and I cant make this work for some reason:
```

(byebug) json_response.keys
["id", "number", "user_id", "full_name", "email", "phone", "completed_at", "display_total", "edit_path", "state", "payment_state", "shipment_state", "payments_path", "ship_path", "ready_to_ship", "created_at", "distributor_name", "special_instructions", "payment_capture_path", "item_total", "adjustment_total", "payment_total", "total", "distributor", "order_cycle", "shipping_method", "ship_address", "bill_address", "line_items", "adjustments", "payments"]
(byebug) order_detailed_attributes
["number"]
(byebug) expect(json_response.keys).to include order_detailed_attributes

  1) Api::OrdersController#show as distributor owner returns an order with all required fields
     Failure/Error: expect(json_response.attributes).to include(order_detailed_attributes)
       expected ["id", "number", "user_id", "full_name", "email", "phone", "completed_at", "display_total", "edit_pat..._cycle", "shipping_method", "ship_address", "bill_address", "line_items", "adjustments", "payments"] to include ["number"]
     # (byebug):1:in `expect_detailed_attributes_to_be_present'
     # /Users/luisramos/.rvm/gems/ruby-2.1.5@openfoodnetwork/gems/byebug-9.0.6/lib/byebug/helpers/eval.rb:59:in `eval'

```

and no luck with have_attributes :-(

Reply:
we need to compare the model attributes hash against the key/values we want and then I don't think we'll need these methods we abstracted. I'll try to give it a crack in your branch.